File: Difficulty: Medium/Min Swaps to Group 1s/min-swaps-to-group-1s.cpp
class Solution {
  public:
    int minSwaps(vector<int>& arr) {
        // code here
        int n=arr.size(), ones=0;
        for(int x:arr) if(x) ones++;
        if(ones==0) return -1;
        int swaps=1e9, left=0, right=0, zeros=0;
        while(right<n){
            if(arr[right]==0) zeros++;
            if(right-left+1==ones){
                swaps=min(swaps,zeros);
                if(arr[left]==0) zeros--;
                left++;
            }
            right++;
        }
        return swaps;
    }
};

File: Difficulty: Medium/Min Swaps to Group 1s/min-swaps-to-group-1s.java
class Solution {
    public int minSwaps(int[] arr) {
        // code here
        int n = arr.length;
        int one = 0;
        for(int i=0;i<n;i++){
            one += arr[i];
        }
        if(one==0) return -1;

        int sum = 0;
        for(int i=0;i<one;i++){
            sum += arr[i];
        }

        int ans = one - sum;
        for(int i=one;i<n;i++){
            sum = sum - arr[i-one] + arr[i];
            int temp = one - sum;
            ans = Math.min(ans, temp);
        }
        return ans;
    }
}
